### ISSUE TITLE
confusion matrix

### ISSUE THREAD

**Issue by Ellohiye** [ultralytics/ultralytics#22609]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I found that the confusion matrix generated after using `detect val` does not correspond to the actual image detection results. For example, if there are 5 instances of a certain target in an image, the confusion matrix shows that 2 instances of that target are detected, and 3 instances are identified as background. However, when I actually detected the target, I found that 3 instances of that target were detected, mistakenly identifying 1 instance of another target.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Ellohiye, thank you for opening this issue and for your interest in Ultralytics 🚀! This is an automated response to help get things moving quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report related to the confusion matrix, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For example:
- exact command(s) you ran (e.g., `yolo detect val ...`)
- the model used and version of `ultralytics`
- a small sample of your dataset with labels that reproduces the issue
- the generated confusion matrix file and corresponding validation logs
- screenshots comparing the confusion matrix counts vs. actual detections you observed

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training/validation logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). This helps us understand class mappings, NMS thresholds, and any potential label mismatches that could influence confusion matrix outputs.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Confusion matrix is generated at `conf=0.25` and with multi-label prediction enabled. Default prediction doesn't use multi-label.

**Comment by Y-T-G** [comment]
> However, when I actually detected the target, I found that 3 instances of that target were detected, mistakenly identifying 1 instance of another target.

That's 4 instances. You said there are 5 instances. So what happened to the 5th one?

**Comment by Ellohiye** [comment]
The fifth one was not identified.

**Comment by Ellohiye** [comment]
The detection results I observed using `detect val` did not show any multi-label predictions.

**Comment by Y-T-G** [comment]
You can run validation with `visualize=True` and it will save the FP, FN images which are derived from the confusion matrix.

**Comment by Ellohiye** [comment]
My question is whether `detect_val` will generate multi-label predictions, rather than `detect_predict`.

**Comment by Y-T-G** [comment]
Validation uses multi-label prediction. If you want to see the exact predictions produced by validation, you can use `visualize=True` to save the output image.

**Comment by Ellohiye** [comment]
What does multi-label prediction mean? Which parameter should be adjusted? Should bs=1?

**Comment by Ellohiye** [comment]
Since detect val also uses multi-label prediction, why is the result different from the confusion matrix? This is the question I asked at the beginning.

**Comment by glenn-jocher** [comment]
Thanks for the follow-up—this mismatch is expected: the confusion matrix matches GT using IoU=0.45 and filters detections at `conf=0.25`, so a box you “see” as a wrong-class hit may be counted as FN (background) for the true class plus an FP for the predicted class if IoU≤0.45; to align what you see with the matrix, run validation with an explicit `conf` and inspect the saved FP/FN overlays derived from the same matches:

```bash
yolo detect val model=your.pt data=your.yaml conf=0.25 plots=True visualize=True
```

Note that “multi‑label” here means a single box can be assigned to multiple classes if their scores exceed `conf`, but the rendered images still show one label per box; details are in the ConfusionMatrix API (see the `conf=0.25` and `iou_thres=0.45` defaults) and validation arguments at [ConfusionMatrix reference](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.ConfusionMatrix.process_batch) and [Model Validation guide](https://docs.ultralytics.com/modes/val/). If it still looks off after setting `conf`, please share one example image with its label file and the saved visualization for us to reproduce.

**Comment by Ellohiye** [comment]
When using `detect val`, I specified an IOU threshold of 0.7. Shouldn't the confusion matrix also use 0.7? So I'm still unclear about the difference between the two.

Does `detect predict` not include multi-label prediction? `detect val` only retains bounding boxes for one class. Is there a parameter that can be modified? I want to retain bounding boxes for multiple classes during validation.

**Comment by glenn-jocher** [comment]
Thanks for the clarifications. Two points:

- The `iou=0.7` you pass to `detect val` controls NMS only; the confusion-matrix matching IoU is a separate, fixed 0.45 internally and isn’t tied to that arg. See what the `iou` flag does in the validation arguments at the Configuration page’s validation settings section for NMS IoU under [Validation settings](https://docs.ultralytics.com/usage/cfg/#validation-settings), and note validation’s postprocess uses multi‑label NMS as shown in [postprocess in detection validation](https://docs.ultralytics.com/reference/models/yolo/detect/val/#postprocess).

- `detect predict` uses single‑label per box by default; validation runs NMS with `multi_label=True`, but the rendered images still display one label per box. To keep multiple class boxes during validation, lower `conf` so additional class scores pass the threshold and optionally save raw predictions for inspection:

```bash
yolo detect val model=your.pt data=your.yaml conf=0.05 iou=0.7 save_txt=True save_conf=True plots=True visualize=True
```

If a specific example still disagrees with the matrix, please attach one image + its label file and the saved `labels/*.txt` and FP/FN visualization so we can reproduce exactly.

**Comment by Ellohiye** [comment]
I believe that regardless of the confidence threshold, only one class of bounding box should be retained, because the model might confuse the classes. It should be possible that the same object can have two predicted classes. Am I misunderstanding the NMS process?

**Comment by glenn-jocher** [comment]
You’re close: in `detect val` NMS is class-aware and runs with multi‑label enabled, so two boxes with the same coordinates but different class IDs can both survive if their scores exceed `conf`; the confusion matrix then matches only one box to the GT and counts the others as FPs, which is why it can differ from what you “see.” No need to change `bs`. For how class-aware NMS behaves, see [Non‑Maximum Suppression (NMS) Explained](https://www.ultralytics.com/glossary/non-maximum-suppression-nms), and for where those exact predictions feed into the matrix, see [`update_metrics` in detection validation](https://docs.ultralytics.com/reference/models/yolo/detect/val/#update_metrics).

**Comment by Ellohiye** [comment]
But why did you say in your previous answer that multi-label detection means: one target can only have one predicted bounding box for one category?

Could you please tell me where the confidence threshold parameter for the confusion matrix is ​​set?

**Comment by glenn-jocher** [comment]
Thanks for the follow-up—my earlier note was about rendering: with multi-label NMS the same box can be emitted for multiple classes, but the viewer draws one label per box; the confusion matrix then matches each GT to at most one detection and counts any extra class boxes as FPs. The confusion matrix confidence comes from the `conf` you pass to `detect val`; if left at the validation default (0.001 for detect, 0.01 for OBB) it’s internally reset to 0.25 inside `ConfusionMatrix.process_batch`. Set it explicitly to override, for example:

```bash
yolo detect val model=your.pt data=your.yaml conf=0.10 plots=True visualize=True
```

For where this happens, see the ConfusionMatrix implementation in the repository under [ConfusionMatrix.process_batch source](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/metrics.py), and for what validation saves (including confusion matrices) see the [Performance metrics deep dive](https://docs.ultralytics.com/guides/yolo-performance-metrics/).

**Comment by Ellohiye** [comment]
Thank you very much for your reply. So what operation are you referring to with this rendering? Is it rendering the image output by predict and val?

**Comment by glenn-jocher** [comment]
By “rendering” I mean the images with boxes/labels that YOLO saves: predict draws its NMS detections on the inputs, while val draws the per‑batch prediction montage (files like `val_batchX_pred.jpg`, see [plot_predictions reference](https://docs.ultralytics.com/reference/models/yolo/detect/val/#plot_predictions)) and, if `visualize=True`, the FP/TP/FN/GT grids derived from the confusion‑matrix matches (saved under `visualizations/`, see [ConfusionMatrix.plot_matches](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.ConfusionMatrix.plot_matches)).

**Comment by Ellohiye** [comment]
After rendering the image, don't val and predict output the saved image? Do these two operations mean that a detection frame can have multiple category labels?

**Comment by glenn-jocher** [comment]
Both `detect predict` and `detect val` save rendered images, but in those images each drawn box always has a single class label; with multi‑label NMS in validation, the same physical object can instead produce several overlapping boxes at the same location, each with a different class, so it’s “multi‑label” via multiple boxes rather than one box showing multiple labels. For a side‑by‑side comparison of how prediction images are produced and saved, see the [Predict mode docs](https://docs.ultralytics.com/modes/predict/).

**Comment by Ellohiye** [comment]
Of course, only one category label can be displayed in one box. Could it be that my previous statement was wrong, which caused you to misunderstand my question? I'm so sorry!
I think the same target may be detected as different categories in the same position after the model is detected, so I want to finally confirm that val and predict are really a target that can only be recognized as one category?

**Comment by glenn-jocher** [comment]
No worries at all, and thanks for sticking with the details here.  

In both `detect predict` and `detect val`, every **detection box** has exactly one class label; there is no single box with multiple category labels. The difference is that validation calls `non_max_suppression(..., multi_label=True, ...)`, so the same physical object can produce several overlapping boxes at the same position with different classes, while prediction uses the default `multi_label=False`, so each candidate box is reduced to its best class only. You can see this behavior and the `multi_label` flag described in the [`non_max_suppression` utility reference](https://docs.ultralytics.com/reference/utils/nms/#ultralytics.utils.nms.non_max_suppression).

**Comment by Ellohiye** [comment]
I noticed that when other models predict images, the same target can have multiple category labels.

**Comment by glenn-jocher** [comment]
Good observation: some frameworks truly attach multiple labels to one box, but Ultralytics YOLO’s detection head is trained with one-hot (single-class) targets per box, so even with “multi-label” NMS in `detect val` you only ever get separate overlapping boxes for different classes at the same location, not a single box that carries multiple category labels.

**Comment by Ellohiye** [comment]
What I am talking about is that when other models use predict, a target can have multiple categories of detection boxes. It has never been said that a detection box has multiple categories.

**Comment by glenn-jocher** [comment]
Got it, we’re on the same page: in Ultralytics YOLO both `detect predict` and `detect val` can produce multiple overlapping boxes with different class IDs for the same physical object; the difference is just that validation enables `multi_label=True` so one candidate can spawn several class-specific boxes, while `predict` calls `non_max_suppression` without `multi_label` (so each candidate keeps only its best class), as shown in the [`DetectionPredictor.postprocess` docs](https://docs.ultralytics.com/reference/models/yolo/detect/predict/#postprocess).
